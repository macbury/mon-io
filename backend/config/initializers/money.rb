# encoding : utf-8
require 'currency_rate'
require_relative '../../app/services/service'
require_relative '../../app/services/exchange_rate_service'

Money.locale_backend = :i18n

CurrencyRate.configure do |config|
  # CurrencyRate::FileStorage configuration has only 'path' attribute
  config.file_storage[:path] = Rails.root.join('storage')

  if ENV.fetch('FIXER_API_KEY', '').present?
    config.api_keys['FixerAdapter'] = ENV.fetch('FIXER_API_KEY')
    # Fiat exchange adapters to fetch
    config.fiat_adapters = ['Fixer']
  end

  # Cryptocurrency exchange adapters to fetch
  config.crypto_adapters = ['Coinbase']
end

MoneyRails.configure do |config|
  # To set the default currency
  #
  config.default_currency = ENV.fetch('MAIN_CURRENCY')
  config.default_bank = Money::Bank::VariableExchange.new(ExchangeRateService)

  # Add exchange rates to current money bank object.
  # (The conversion rate refers to one direction only)
  #
  # Example:
  # config.add_rate "USD", "CAD", 1.24515
  # config.add_rate "CAD", "USD", 0.803115

  # To handle the inclusion of validations for monetized fields
  # The default value is true
  #
  # config.include_validations = true

  # Default ActiveRecord migration configuration values for columns:
  #
  # config.amount_column = { prefix: '',           # column name prefix
  #                          postfix: '_cents',    # column name  postfix
  #                          column_name: nil,     # full column name (overrides prefix, postfix and accessor name)
  #                          type: :integer,       # column type
  #                          present: true,        # column will be created
  #                          null: false,          # other options will be treated as column options
  #                          default: 0
  #                        }
  #
  # config.currency_column = { prefix: '',
  #                            postfix: '_currency',
  #                            column_name: nil,
  #                            type: :string,
  #                            present: true,
  #                            null: false,
  #                            default: 'USD'
  #                          }

  # Register a custom currency
  #
  Money::Currency.unregister('XAU')
  Money::Currency.unregister('XAG')

  config.register_currency = {
    priority: 100,
    iso_code: 'XAU',
    name: 'Gold (Troy Ounce)',
    symbol: 'oz t',
    disambiguate_symbol: 'XAU',
    alternate_symbols: [],
    subunit: 'oz',
    subunit_to_unit: 100,
    symbol_first: false,
    format: '%n %u',
    html_entity: '',
    decimal_mark: '.',
    thousands_separator: ',',
    iso_numeric: 959,
    smallest_denomination: 1
  }

  config.register_currency = {
    priority: 100,
    iso_code: 'XAG',
    name: 'Silver (Troy Ounce)',
    symbol: 'oz t',
    disambiguate_symbol: 'XAG',
    alternate_symbols: [],
    subunit: 'oz',
    subunit_to_unit: 100,
    symbol_first: false,
    format: '%n %u',
    html_entity: '',
    decimal_mark: '.',
    thousands_separator: ',',
    iso_numeric: 959,
    smallest_denomination: 1
  }

  config.register_currency = {
    priority:            1,
    iso_code:            "ETH",
    name:                "Ethereum",
    symbol:              "Ξ",
    symbol_first:        false,
    subunit:             "WEI",
    subunit_to_unit:     100000000,
    thousands_separator: ".",
    decimal_mark:        ","
  }

  config.register_currency = {
    priority:            1,
    iso_code:            "GAU",
    name:                "Gold Gram",
    symbol:              "GAU",
    symbol_first:        false,
    subunit:             "GAU",
    subunit_to_unit:     1 ,
    thousands_separator: ".",
    decimal_mark:        ","
  }

  config.register_currency = {
    priority:            1,
    iso_code:            "GAU",
    name:                "Gold Gram",
    symbol:              "GAU",
    symbol_first:        false,
    subunit:             "GAU",
    subunit_to_unit:     1 ,
    thousands_separator: ".",
    decimal_mark:        ","
  }

  config.register_currency = {
    priority:            1,
    iso_code:            "GAG",
    name:                "Silver Gram",
    symbol:              "GAG",
    symbol_first:        false,
    subunit:             "GAG",
    subunit_to_unit:     1 ,
    thousands_separator: ".",
    decimal_mark:        ","
  }

  # Specify a rounding mode
  # Any one of:
  #
  # BigDecimal::ROUND_UP,
  # BigDecimal::ROUND_DOWN,
  # BigDecimal::ROUND_HALF_UP,
  # BigDecimal::ROUND_HALF_DOWN,
  # BigDecimal::ROUND_HALF_EVEN,
  # BigDecimal::ROUND_CEILING,
  # BigDecimal::ROUND_FLOOR
  #
  # set to BigDecimal::ROUND_HALF_EVEN by default
  #
  config.rounding_mode = BigDecimal::ROUND_HALF_UP

  # Set default money format globally.
  # Default value is nil meaning "ignore this option".
  # Example:
  #
  # config.default_format = {
  #   no_cents_if_whole: nil,
  #   symbol: nil,
  #   sign_before_symbol: nil
  # }

  # If you would like to use I18n localization (formatting depends on the
  # locale):
  config.locale_backend = :i18n
  #
  # Example (using default localization from rails-i18n):
  #
  # I18n.locale = :en
  # Money.new(10_000_00, 'USD').format # => $10,000.00
  # I18n.locale = :es
  # Money.new(10_000_00, 'USD').format # => $10.000,00
  #
  # For the legacy behaviour of "per currency" localization (formatting depends
  # only on currency):
  # config.locale_backend = :currency
  #
  # Example:
  # Money.new(10_000_00, 'USD').format # => $10,000.00
  # Money.new(10_000_00, 'EUR').format # => €10.000,00
  #
  # In case you don't need localization and would like to use default values
  # (can be redefined using config.default_format):
  # config.locale_backend = nil

  # Set default raise_error_on_money_parsing option
  # It will be raise error if assigned different currency
  # The default value is false
  #
  # Example:
  # config.raise_error_on_money_parsing = false
end
