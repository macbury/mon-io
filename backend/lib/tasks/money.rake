namespace :money do
  desc "Refresh rates when called"
  task refresh_rates: :environment do
    Currency::SyncExchangeRate.call
  end
end