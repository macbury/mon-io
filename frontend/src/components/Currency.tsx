import React from 'react'
import { Money } from '../api/graphql'
import { useMoneyParts } from '../helpers/currency'
import { Text } from 'react-native-paper'
import styled, { DefaultTheme } from 'styled-components/native'

const Container = styled.View`
  display: flex;
  flex-direction: row;
`

interface IMoneyPart {
  expense?: boolean | number
  theme: DefaultTheme
  size: number;
}

function getTextExpenseTextColor({ theme, expense } : IMoneyPart) {
  return expense ? theme.expenseColor : theme.colors.text
}

const Dollars = styled(Text)`
  font-weight: bold;
  font-size: ${(prop : IMoneyPart) => prop.size}px;
  color: ${getTextExpenseTextColor};
`

const Cents = styled(Text)`
  font-size: ${(prop : IMoneyPart) => prop.size - 6}px;
  font-weight: bold;
  color: ${getTextExpenseTextColor};
`

const CurrencySymbol = styled(Text)`
  font-weight: bold;
  font-size: ${(prop : IMoneyPart) => prop.size}px;
  margin-left: 10px;
  opacity: 0.3;
`

export interface ICurrencyProps {
  amount: Money
  size: number
  showCurrency?: boolean
}

export default function Currency({ amount, size, showCurrency, ...props } : ICurrencyProps) {
  const [dollars, cents, currency] = useMoneyParts(amount)
  const dollarsWithSpaces = Math.abs(dollars as number).toLocaleString('en').split(/\-/).join(' ')
  const expense = dollars < 0
  const sign = expense ? '- ' : ''

  return (
    <Container {...props}>
      <Dollars expense={expense ? 1 : 0} size={size}>{sign}{dollarsWithSpaces}.</Dollars>
      <Cents expense={expense ? 1 : 0} size={size}>{cents}</Cents>
      <CurrencySymbol size={size}>{currency}</CurrencySymbol>
    </Container>
  )
}