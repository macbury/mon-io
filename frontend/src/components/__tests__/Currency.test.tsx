import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import Currency, { ICurrencyProps } from '../Currency'

import { money } from '../../test/factories/money'

it('renders correctly', () => {
  const props = {
    amount: money.build(),
    size: 32,
    showCurrency: true
  }

  const component = shallow(<Currency {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('renders bitcoin', () => {
  const props = {
    amount: {
      "cents": -230520,
      "currency": {
        "id": "BTC",
        "name": "Bitcoin",
        "symbol":"â‚¿",
        "isoCode":"BTC",
        "subunitToUnit":100000000
      },
      "__typename":"Money"
    },
    size: 32,
    showCurrency: true
  }

  const component = shallow(<Currency {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('do not throw error for empty amount', () => {
  const props = {
    amount: null,
    size: 32,
    showCurrency: true
  }

  const component = shallow(<Currency {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('do not throw error for empty currency', () => {
  const props : ICurrencyProps = {
    amount: {
      cents: 100,
      currency: null
    },
    size: 32,
    showCurrency: true
  }

  const component = shallow(<Currency {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})