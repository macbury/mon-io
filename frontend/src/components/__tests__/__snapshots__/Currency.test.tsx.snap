// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`do not throw error for empty amount 1`] = `
<Styled(View)>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0
    .
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    size={32}
  >
    EUR
  </Styled(withTheme(Text))>
</Styled(View)>
`;

exports[`do not throw error for empty currency 1`] = `
<Styled(View)>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0
    .
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    size={32}
  >
    EUR
  </Styled(withTheme(Text))>
</Styled(View)>
`;

exports[`renders bitcoin 1`] = `
<Styled(View)>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0
    .
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    0023052
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    size={32}
  >
    ₿
  </Styled(withTheme(Text))>
</Styled(View)>
`;

exports[`renders correctly 1`] = `
<Styled(View)>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    10
    .
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    expense={0}
    size={32}
  >
    00
  </Styled(withTheme(Text))>
  <Styled(withTheme(Text))
    size={32}
  >
    zł
  </Styled(withTheme(Text))>
</Styled(View)>
`;
